File: cuelens/app/src/test/java/de/eachandevery/cuelens/prestudy/ParticipantIdentifierTest.kt
package de.eachandevery.cuelens.prestudy

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class ParticipantIdentifierTest {
    @Test
    fun acceptsProlificIdsAndPreservesCase() {
        val cases = listOf(
            "AbCdEf1234567890GhIjKlMn" to "AbCdEf1234567890GhIjKlMn",
            " \tAbCdEf1234567890GhIjKlMn\r\n" to "AbCdEf1234567890GhIjKlMn",
            "123456789012345678901234" to "123456789012345678901234"
        )

        cases.forEach { (input, expected) ->
            val identifier = ParticipantIdentifier.parse(input)

            assertTrue(identifier is ParticipantIdentifier.ProlificId)
            assertEquals(expected, identifier?.value)
        }
    }

    @Test
    fun acceptsEmailWithoutChangingItsSpelling() {
        val identifier = ParticipantIdentifier.parse("  Participant@Example.ORG  ")

        assertTrue(identifier is ParticipantIdentifier.DirectEmail)
        assertEquals("Participant@Example.ORG", identifier?.value)
    }

    @Test
    fun rejectsSameInvalidValuesAsServerParser() {
        val invalidIdentifiers = listOf(
            "",
            "   ",
            "AbCdEf1234567890GhIjKlM",
            "AbCdEf1234567890GhIjKlMnO",
            "AbCdEf123456 7890GhIjKlM",
            "AbCdEf1234567890GhIjKlM-",
            "AbCdEf1234567890GhIjKlMö",
            "AbCdEf123456\n890GhIjKlMn",
            "participant@example"
        )

        invalidIdentifiers.forEach { input ->
            assertNull("Expected rejection for: $input", ParticipantIdentifier.parse(input))
        }
    }
}

File: cuelens.each-and-every.de/tests/ParticipantIdentifierTest.php
<?php
declare(strict_types=1);

use PHPUnit\Framework\Attributes\DataProvider;
use PHPUnit\Framework\TestCase;

require_once dirname(__DIR__) . '/lib/participant-identifier.php';

final class ParticipantIdentifierTest extends TestCase
{
    #[DataProvider('validProlificIdProvider')]
    public function testAcceptsProlificIdsAndPreservesCase(string $input, string $expected): void
    {
        $identifier = ParticipantIdentifier::parse($input);

        self::assertSame(ParticipantIdentifier::PROLIFIC, $identifier->channel());
        self::assertSame($expected, $identifier->value());
        self::assertSame($expected, $identifier->activationValue());
        self::assertSame($expected, $identifier->prolificId());
        self::assertNull($identifier->email());
    }

    /** @return iterable<string, array{string, string}> */
    public static function validProlificIdProvider(): iterable
    {
        yield 'mixed case' => ['AbCdEf1234567890GhIjKlMn', 'AbCdEf1234567890GhIjKlMn'];
        yield 'surrounding whitespace' => [" \tAbCdEf1234567890GhIjKlMn\r\n", 'AbCdEf1234567890GhIjKlMn'];
        yield 'digits only' => ['123456789012345678901234', '123456789012345678901234'];
    }

    public function testAcceptsEmailWithoutChangingItsStoredSpelling(): void
    {
        $identifier = ParticipantIdentifier::parse('  Participant@Example.ORG  ');

        self::assertSame(ParticipantIdentifier::DIRECT, $identifier->channel());
        self::assertSame('Participant@Example.ORG', $identifier->value());
        self::assertSame('participant@example.org', $identifier->activationValue());
        self::assertSame('Participant@Example.ORG', $identifier->email());
        self::assertNull($identifier->prolificId());
    }

    #[DataProvider('invalidIdentifierProvider')]
    public function testRejectsInvalidIdentifiers(string $input): void
    {
        $this->expectException(InvalidArgumentException::class);

        ParticipantIdentifier::parse($input);
    }

    /** @return iterable<string, array{string}> */
    public static function invalidIdentifierProvider(): iterable
    {
        yield 'empty' => [''];
        yield 'whitespace' => ['   '];
        yield '23 characters' => ['AbCdEf1234567890GhIjKlM'];
        yield '25 characters' => ['AbCdEf1234567890GhIjKlMnO'];
        yield 'internal whitespace' => ['AbCdEf123456 7890GhIjKlM'];
        yield 'punctuation' => ['AbCdEf1234567890GhIjKlM-'];
        yield 'unicode look alike' => ['AbCdEf1234567890GhIjKlMö'];
        yield 'line break inside' => ["AbCdEf123456\n890GhIjKlMn"];
        yield 'malformed email' => ['participant@example'];
    }
}
